File: tests/tensors.test.ts
import { expect, test, describe, beforeEach } from 'vitest';
import { device, Tensors, DType, isAvailable } from '@/webgpu-tensors';
import { Device } from 'packages/core/src/types';

const testTensors = (tensors: Tensors, device: Device) => {
  describe('Tensors', () => {
    test('Get Tensors Type', () => {
      expect(tensors.device).toBe(device);
    });
  });
  describe('Tensor Operations', () => {
    beforeEach(async () => {
      await tensors.init();
    });

    test('Create empty tensor', () => {
      const shape = [2, 3];
      const tensor = tensors.empty(shape);
      expect(tensor).toBeInstanceOf(Object);
      expect(tensor.shape.data).toEqual(shape);
      expect(tensor.dtype).toBe(DType.Float32);
    });

    test('Create ones tensor', async () => {
      const shape = [2, 2];
      const tensor = tensors.ones(shape);
      expect(tensor).toBeInstanceOf(Object);
      expect(tensor.shape.data).toEqual(shape);
      const data = await tensors.clone(tensor).readArray<number>();
      expect(data).toEqual([
        [1, 1],
        [1, 1],
      ]);
    });

    test('Create zeros tensor', async () => {
      const shape = [2, 3];
      const tensor = tensors.zeros(shape);
      expect(tensor).toBeInstanceOf(Object);
      expect(tensor.shape.data).toEqual(shape);
      const data = await tensors.clone(tensor).readArray<number>();
      expect(data).toEqual([
        [0, 0, 0],
        [0, 0, 0],
      ]);
    });

    test('Create rand tensor', async () => {
      const shape = [2, 3];
      const tensor = tensors.rand(shape);
      expect(tensor).toBeInstanceOf(Object);
      expect(tensor.shape.data).toEqual(shape);
      const data = await tensors.clone(tensor).readArray<number>();
      expect((data as number[]).flat().every((v) => v >= 0 && v < 1)).toBe(true);
    });

    test('Create randn tensor', async () => {
      const shape = [1000];
      const tensor = tensors.randn(shape);
      expect(tensor).toBeInstanceOf(Object);
      expect(tensor.shape.data).toEqual(shape);
      const data = await tensors.clone(tensor).readArray<number>();
      const mean = (data as number[]).reduce((sum, val) => sum + val, 0) / data.length;
      const variance =
        (data as number[]).reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / data.length;
      expect(mean).toBeCloseTo(0, 1);
      expect(Math.sqrt(variance)).toBeCloseTo(1, 1);
    });

    test('Matrix multiplication', async () => {
      const a = tensors.tensor([
        [1, 2],
        [3, 4],
      ]);
      const b = tensors.tensor([
        [5, 6],
        [7, 8],
      ]);
      const result = tensors.matmul(a, b);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([
        [19, 22],
        [43, 50],
      ]);
    });

    test('Subtraction', async () => {
      const a = tensors.tensor([1, 2, 3]);
      const b = tensors.tensor([4, 5, 6]);
      const result = tensors.sub(a, b);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([-3, -3, -3]);
    });

    test('Power', async () => {
      const a = tensors.tensor([1, 2, 3]);
      const result = tensors.pow(a, 2);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([1, 4, 9]);
    });

    test('Multiplication', async () => {
      const a = tensors.tensor([1, 2, 3]);
      const b = tensors.tensor([4, 5, 6]);
      const result = tensors.mul(a, b);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([4, 10, 18]);
    });

    test('Greater than', async () => {
      const a = tensors.tensor([1, 2, 3]);
      const result = tensors.gt(a, 2);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([0, 0, 1]);
    });

    test('Transpose', async () => {
      const a = tensors.tensor([
        [1, 2, 3],
        [4, 5, 6],
      ]);
      const result = tensors.transpose(a);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([
        [1, 4],
        [2, 5],
        [3, 6],
      ]);
    });

    test('Maximum', async () => {
      const a = tensors.tensor([-1, 0, 1]);
      const result = tensors.maximum(a, 0);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([0, 0, 1]);
    });

    test('ReLU activation', async () => {
      const input = tensors.tensor([-2, -1, 0, 1, 2]);
      const result = tensors.relu(input);
      const data = await tensors.clone(result).readArray<number>();
      expect(data).toEqual([0, 0, 0, 1, 2]);
    });

    test('Mean', async () => {
      const a = tensors.tensor([1, 2, 3, 4, 5]);
      const result = tensors.mean(a);
      const data = await tensors.clone(result).readArray<number>();
      expect(data[0]).toBeCloseTo(3);
    });

    test('Sigmoid', async () => {
      const a = tensors.tensor([-2, -1, 0, 1, 2]);
      const result = tensors.sigmoid(a);
      const data = await tensors.clone(result).readArray<number>();
      expect((data as number[]).map((v) => Number(v.toFixed(4)))).toEqual([
        0.1192, 0.2689, 0.5, 0.7311, 0.8808,
      ]);
    });

    test('Max', async () => {
      const a = tensors.tensor([1, 5, 3, 2, 4]);
      const result = tensors.max(a);
      const data = await tensors.clone(result).readArray<number>();
      expect(data[0]).toBe(5);
    });
  });
};

describe('WebGPUTensors', () => {
  if (isAvailable(Device.GPU)) {
    testTensors(device(Device.GPU), Device.GPU);
  } else {
    test('No WebGPU Support', async () => {
      expect(isAvailable(Device.GPU)).toBe(false);
    });
  }
});

describe('JSTensors', () => {
  testTensors(device(Device.CPU), Device.CPU);
});

describe('WASMTensors', () => {
  testTensors(device(Device.WASM), Device.WASM);
});

File: tests/tensors_test.py
import unittest
import torch

class TestPyTorchTensors(unittest.TestCase):
    def test_create_empty_tensor(self):
        shape = [2, 3]
        tensor = torch.empty(shape)
        self.assertIsInstance(tensor, torch.Tensor)
        self.assertEqual(list(tensor.shape), shape)
        self.assertEqual(tensor.dtype, torch.float32)

    def test_create_ones_tensor(self):
        shape = [2, 2]
        tensor = torch.ones(shape)
        self.assertIsInstance(tensor, torch.Tensor)
        self.assertEqual(list(tensor.shape), shape)
        torch.testing.assert_close(tensor, torch.ones(shape))

    def test_create_zeros_tensor(self):
        shape = [2, 3]
        tensor = torch.zeros(shape)
        self.assertIsInstance(tensor, torch.Tensor)
        self.assertEqual(list(tensor.shape), shape)
        torch.testing.assert_close(tensor, torch.zeros(shape))

    def test_create_rand_tensor(self):
        shape = [2, 3]
        tensor = torch.rand(shape)
        self.assertIsInstance(tensor, torch.Tensor)
        self.assertEqual(list(tensor.shape), shape)
        self.assertTrue((tensor >= 0).all() and (tensor < 1).all())

    def test_create_randn_tensor(self):
        shape = [1000]  # Large shape for better statistical properties
        tensor = torch.randn(shape)
        self.assertIsInstance(tensor, torch.Tensor)
        self.assertEqual(list(tensor.shape), shape)
        self.assertAlmostEqual(tensor.mean().item(), 0, delta=0.1)
        self.assertAlmostEqual(tensor.std().item(), 1, delta=0.1)

    def test_matrix_multiplication(self):
        a = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32)
        b = torch.tensor([[5, 6], [7, 8]], dtype=torch.float32)
        result = torch.matmul(a, b)
        expected = torch.tensor([[19, 22], [43, 50]], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_subtraction(self):
        a = torch.tensor([1, 2, 3], dtype=torch.float32)
        b = torch.tensor([4, 5, 6], dtype=torch.float32)
        result = torch.sub(a, b)
        expected = torch.tensor([-3, -3, -3], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_power(self):
        a = torch.tensor([1, 2, 3], dtype=torch.float32)
        result = torch.pow(a, 2)
        expected = torch.tensor([1, 4, 9], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_multiplication(self):
        a = torch.tensor([1, 2, 3], dtype=torch.float32)
        b = torch.tensor([4, 5, 6], dtype=torch.float32)
        result = torch.mul(a, b)
        expected = torch.tensor([4, 10, 18], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_greater_than(self):
        a = torch.tensor([1, 2, 3], dtype=torch.float32)
        result = a > 2
        expected = torch.tensor([False, False, True])
        torch.testing.assert_close(result, expected)

    def test_transpose(self):
        a = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float32)
        result = torch.transpose(a, 0, 1)
        expected = torch.tensor([[1, 4], [2, 5], [3, 6]], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_maximum(self):
        a = torch.tensor([-1, 0, 1], dtype=torch.float32)
        result = torch.maximum(a, torch.tensor(0))
        expected = torch.tensor([0, 0, 1], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_relu_activation(self):
        input_tensor = torch.tensor([-2, -1, 0, 1, 2], dtype=torch.float32)
        result = torch.relu(input_tensor)
        expected = torch.tensor([0, 0, 0, 1, 2], dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_mean(self):
        a = torch.tensor([1, 2, 3, 4, 5], dtype=torch.float32)
        result = torch.mean(a)
        expected = torch.tensor(3, dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_softmax(self):
        a = torch.tensor([1, 2, 3], dtype=torch.float32)
        result = torch.softmax(a, dim=0)
        expected = torch.tensor([0.0900, 0.2447, 0.6652], dtype=torch.float32)
        torch.testing.assert_close(result, expected, atol=1e-4, rtol=1e-4)

    def test_sigmoid(self):
        a = torch.tensor([-2, -1, 0, 1, 2], dtype=torch.float32)
        result = torch.sigmoid(a)
        expected = torch.tensor([0.1192, 0.2689, 0.5000, 0.7311, 0.8808], dtype=torch.float32)
        torch.testing.assert_close(result, expected, atol=1e-4, rtol=1e-4)

    def test_max(self):
        a = torch.tensor([1, 5, 3, 2, 4], dtype=torch.float32)
        result = torch.max(a)
        expected = torch.tensor(5, dtype=torch.float32)
        torch.testing.assert_close(result, expected)

    def test_sum(self):
        a = torch.tensor([1, 2, 3, 4, 5], dtype=torch.float32)
        result = torch.sum(a)
        expected = torch.tensor(15, dtype=torch.float32)
        torch.testing.assert_close(result, expected)

if __name__ == '__main__':
    unittest.main()
